source/MRMesh/MRPointsComponents.cpp

```diff
@@ -0,0 +1,100 @@
+#include "MRPointsComponents.h"
+#include "MRPointCloud.h"
+#include "MRBitSet.h"
+#include "MRTimer.h"
+#include "MRPointsInBall.h"
+#include "MRProgressCallback.h"
+//#include <algorithm>
+
+namespace MR
+{
+
+namespace PointCloudComponents
+{
+
+Expected<std::vector<MR::VertBitSet>> getLargestComponents( const PointCloud& pointCloud, float maxDist, int minSize /*= -1*/, ProgressCallback pc /*= {}*/ )
+{
+    MR_TIMER
+
+    assert( maxDist > 0.f );
+    const auto& validPoints = pointCloud.validPoints;
+    if ( minSize < 0 )
+        minSize = std::max( int( validPoints.count() ) / 100, 1 );
+    ProgressCallback subPc = pc ? subprogress( pc, 0.f, 0.9f ) : pc;
+    auto unionStructsRes = PointCloudComponents::getUnionFindStructureVerts( pointCloud, maxDist, nullptr, subPc );
+    if ( !unionStructsRes.has_value() )
+        return unexpected( unionStructsRes.error() );
+    auto& unionStructs = *unionStructsRes;
+    auto allRoots = unionStructs.roots();
+    HashMap<VertId, int> rootMap;
+    std::vector<VertBitSet> components;
```

Comment:
You still collect all components, but return only part of them

Reply:
reworked

Reply:
reworked